core/src/test/scala/unit/kafka/coordinator/transaction/TransactionCoordinatorTest.scala

```diff
@@ -842,6 +897,43 @@ class TransactionCoordinatorTest {
     EasyMock.verify(transactionManager)
   }
 
+  @Test
+  def shouldNotAcceptSmallerEpochDuringTransactionExpiration(): Unit = {
+    val now = time.milliseconds()
+    val txnMetadata = new TransactionMetadata(transactionalId, producerId, producerId, producerEpoch,
+      RecordBatch.NO_PRODUCER_EPOCH, txnTimeoutMs, Ongoing, partitions, now, now)
+
+    EasyMock.expect(transactionManager.timedOutTransactions())
+      .andReturn(List(TransactionalIdAndProducerIdEpoch(transactionalId, producerId, producerEpoch)))
+    EasyMock.expect(transactionManager.getTransactionState(EasyMock.eq(transactionalId)))
+      .andReturn(Right(Some(CoordinatorEpochAndTxnMetadata(coordinatorEpoch, txnMetadata))))
+
+    val bumpedTxnMetadata = new TransactionMetadata(transactionalId, producerId, producerId, (producerEpoch + 2).toShort,
+      RecordBatch.NO_PRODUCER_EPOCH, txnTimeoutMs, Ongoing, partitions, now, now)
+    EasyMock.expect(transactionManager.getTransactionState(EasyMock.eq(transactionalId)))
+      .andReturn(Right(Some(CoordinatorEpochAndTxnMetadata(coordinatorEpoch, bumpedTxnMetadata))))
+
+    val appender = LogCaptureAppender.createAndRegister()
+    LogCaptureAppender.setClassLoggerLevel(coordinator.getClass, Level.DEBUG)
+    try {
+
+      EasyMock.replay(transactionManager, transactionMarkerChannelManager)
+
+      coordinator.startup(false)
+      time.sleep(TransactionStateManager.DefaultAbortTimedOutTransactionsIntervalMs)
+      scheduler.tick()
+      EasyMock.verify(transactionManager)
+
+      val event = appender.getMessages.find(e => e.getMessage.equals("Rollback of " +
```

Comment:
Tests which capture logging output tend to be brittle. Would it be more direct to call `abortTimedOutTransactions` directly?

Reply:
The tricky thing is `abortTimedOutTransactions` is private and we don't want to expose a function just for testing purpose I guess?